app/models/activity.rb

```diff
@@ -83,14 +80,20 @@ def time_ranges_cache_key
     "Activity#to_time_ranges##{id}##{updated_at.to_f}"
   end
 
-  def time_value(hash)
-    # Store time values on the same day so the duration of each occurrence is correct.
-    Time.zone.local(1, 1, 1, hash[4], hash[5])
-  end
-
   def validate_end_time_after_start_time
     return unless start_time && end_time && end_time <= start_time
 
     errors.add :duration, I18n.t('errors.activity.duration.missing')
   end
+
+  def build_schedule
+    return unless plan
+
+    activity_start_time = plan.start_date.beginning_of_day
+    self.schedule = ScheduleBuilder.call(
+      start_time: activity_start_time,
+      schedule: schedule,
+      minutes_per_week: @seconds_per_week.to_f / 60
+    )
+  end
```

Comment:
yes, the change that makes the difference to the performance of schedule.occurrences_between is to set the start date of the schedule to the start date of the job plan (instead of 01/01/0001). In order to do this I need the plan and that's not present at the point at which the seconds_per_week setter is called.